rclcpp/include/rclcpp/callback_group.hpp

```diff
@@ -89,13 +89,16 @@ class CallbackGroup
    * added to the executor in either case.
    *
    * \param[in] group_type The type of the callback group.
+   * \param[in] get_node_context Lambda to retrieve the node context when
+   *   checking that the creating node is valid and using the guard condition.
    * \param[in] automatically_add_to_executor_with_node A boolean that
    *   determines whether a callback group is automatically added to an executor
    *   with the node with which it is associated.
    */
   RCLCPP_PUBLIC
   explicit CallbackGroup(
     CallbackGroupType group_type,
+    std::function<rclcpp::Context::SharedPtr(void)> get_node_context,
```

Comment:
If we are worried about tick-tocking this class (as we seem to be for `get_notify_guard_condition` below), then don't we need to keep a deprecated version of the constructor that doesn't take in `get_node_context`?

Reply:
I don't know that we have to really worry about tick-tocking as this object is only created via the node `create_callback_group` call.  I started to out of an abundance of caution, but in hindsight it may not really be necessary.

Reply:
Updated to retain both patterns: 87f41bff

If a user constructs the callbackgroup without passing a node context function and also doesn't pass in a context to `get_notify_guard_condition` then an exception will be thrown.